libafl/src/corpus/id_manager.rs

```diff
@@ -0,0 +1,173 @@
+            id: self.id_mappings.len(),
+        };
+        self.active_ids.push(id);
+        self.id_mappings
+            .push(Some(self.active_ids.len().try_into().unwrap()));
+        id
+    }
+
+    /// Invalidate the given [`CorpusId`]. This will cause any future operations and lookups for this [`CorpusId`] to
+    /// fail. If the id was valid, returns the index where it was found;
+    #[must_use]
+    pub(crate) fn remove_id(&mut self, id: CorpusId) -> Option<usize> {
+        // debug_assert!(self.active_ids.is_sorted());
+        self.id_mappings[id.id].take().map(|idx| {
+            let idx = usize::from(idx) - 1;
+            self.active_ids.remove(idx);
+            idx
+        })
+    }
+
+    /// Get the corpus index for the given [`CorpusId`]. Should only ever be called by a [`Corpus`].
+    #[must_use]
+    pub(crate) fn active_index_for(&self, id: CorpusId) -> Option<usize> {
+        // debug_assert!(self.active_ids.is_sorted());
+        self.id_mappings[id.id].map(|idx| usize::from(idx) - 1)
+    }
+
+    /// Get the [`CorpusId`] at the given index. If the index is out of bounds, returns `Err`.
+    pub fn get(&self, idx: usize) -> Result<CorpusId, Error> {
+        self.assert_has_active()?;
+        self.active_ids.get(idx).copied().ok_or_else(|| {
+            Error::illegal_argument(format!(
+                "The given idx {idx} was out of bounds for id_manager (active_ids: {})",
+                self.active_ids.len()
+            ))
+        })
+    }
+
+    /// Returns [`Error::Empty`] if no entries are in `active_ids`, else `Ok`
+    #[inline]
+    fn assert_has_active(&self) -> Result<(), Error> {
+        if self.active_ids.is_empty() {
+            Err(Error::empty("No active_ids in id_manager"))
+        } else {
+            Ok(())
+        }
+    }
+
+    /// Get the id of the first (oldest) active [`CorpusId`].
+    #[must_use]
+    pub fn first_id(&self) -> Option<CorpusId> {
+        self.active_ids.first().copied()
+    }
+
+    /// Gets the "next" (least less old) active [`CorpusId`]. This is like incrementing the index.
+    #[must_use]
+    pub fn find_next(&self, id: CorpusId) -> Option<CorpusId> {
+        // it should ALWAYS be sorted since it only ever gets appended to with larger values (next ids)
+        // debug_assert!(self.active_ids.is_sorted());
+
+        if let Some(idx) = self.id_mappings[id.id] {
+            let idx = usize::from(idx) - 1;
+            // get next entry in the active list
+            if self.active_ids.len() > idx + 1 {
+                return Some(self.active_ids[idx]);
+            }
+        }
+        None
+    }
+
+    /// Returns the current idx for the given corpus id
+    #[must_use]
+    pub fn lookup(&self, corpus_id: CorpusId) -> Option<usize> {
+        self.id_mappings[corpus_id.id].map(|idx| usize::from(idx) - 1)
+    }
+}
+
+/// Return a random entry from the corpus of a given state. This function is mainly for encapsulation to make the borrow
```

Comment:
@andreafioraldi this is the way to get around the borrow checker without a macro